instrumental/drivers/tempcontrollers/covesion.py

```diff
@@ -0,0 +1,230 @@
+from pyvisa.constants import *
+from pyvisa import ResourceManager
+import time
+from datetime import timedelta
+from sys import stdout
+import codecs
+import numpy as np
+from time import sleep
+#import visa
+#from pint import unit
+#from .. import _get_visa_instrument
+#from .. import InstrumentTypeError
+#from ... import visa
+
+# def _instrument(params):
+#     inst = _get_visa_instrument(params)
+#     # Should add a check of instrument type here. Not sure how to do this now,
+#     # since querying '*IDN?' doesn't work.
+#     return OC(inst)
+_INST_PARAMS = ['OC_visa_address','version']
+_INST_CLASSES = ['OC']
+#rm = ResourceManager()
+OC_parity = Parity.none
+OC_baud_rate = 19200
+OC_data_bits = 8
+OC_read_termination = '\r'
+OC_flow_control = 0
+OC_timeout = 500
+OC_status_keys =  ['set point',
+                     'temperature',
+                     'control',
+                     r'output %',
+                     'alarms',
+                     'faults',
+                     'temp ok',
+                     'supply vdc',
+                     'version',
+                     'test cycle',
+                     'test mode',
+                     ]
+rm = ResourceManager()
+def print_statusline(msg: str):
+    last_msg_length = len(print_statusline.last_msg) if hasattr(print_statusline, 'last_msg') else 0
+    print(' ' * last_msg_length, end='\r')
+    print(msg, end='\r')
+    stdout.flush()
+    print_statusline.last_msg = msg
+
+def _open_visa_OC(rm,visa_address):
+    visa_inst = rm.get_instrument(visa_address)
+    visa_inst.parity = OC_parity # = Parity.none
+    visa_inst.baud_rate = OC_baud_rate # = 19200
+    visa_inst.data_bits = OC_data_bits # = 8
+    visa_inst.read_termination = OC_read_termination # = '\r'
+    visa_inst.flow_control = OC_flow_control # = 0
+    visa_inst.timeout = OC_timeout # = 10000
+    visa_inst.clear()
+    return visa_inst
+
+def _check_OC(rm,visa_address,n_tries_max=5):
+    n_tries = 0
+    success = False
+    version = False
+    while not(success) and (n_tries < n_tries_max):
+        try:
+            visa_inst = _open_visa_OC(rm,visa_address)
+            output_raw = visa_inst.ask('\X01J00\X00\XCB')
+            visa_inst.close()
+            success = True
+            vals = output_raw[5:-3].split(';')
+            status_dict = dict(zip(OC_status_keys,vals))
+            version = status_dict['version']
+        except:
+            pass
+        n_tries = n_tries + 1
+    return version
+    #
+    #return dict(zip(self.status_keys,vals))
+
+def list_instruments():
+    instruments = []
+    #rm = ResourceManager()
+    visa_list = rm.list_resources()
+    for addr in visa_list:
+        if addr[0:4]=='ASRL':
+            version = _check_OC(rm,addr)
+            if version:
+                params = ParamSet(OC,OC_visa_address=addr,version=version)
+                instruments.append(params)
+    #rm.close()
+    return instruments
+
+
+class OC(TempController):
```

Comment:
It's probably a good idea to rename this something more detailed, like `CovesionOC`